wfe2/wfe.go

```diff
@@ -1393,59 +1441,61 @@ func (wfe *WebFrontEndImpl) NewOrder(
+	var newOrderRequest struct {
+		Identifiers []core.AcmeIdentifier `json:"identifiers"`
+	}
+	err := json.Unmarshal(body, &newOrderRequest)
 	if err != nil {
-		wfe.sendError(response, logEvent, probs.Malformed("Error unmarshaling order request"), err)
+		wfe.sendError(response, logEvent,
+			probs.Malformed("Unable to unmarshal NewOrder request body"), err)
 		return
 	}
-	// Assuming a properly formatted CSR there should be two four byte SEQUENCE
-	// declarations then a two byte integer declaration which defines the version
-	// of the CSR. If those two bytes (at offset 8 and 9) and equal to 2 and 0
-	// then the CSR was generated by a pre-1.0.2 version of OpenSSL with a client
-	// which didn't explicitly set the version causing the integer to be malformed
-	// and encoding/asn1 will refuse to parse it. If this is the case exit early
-	// with a more useful error message.
-	if len(rawCSR.CSR) >= 10 && rawCSR.CSR[8] == 2 && rawCSR.CSR[9] == 0 {
-		wfe.sendError(
-			response,
-			logEvent,
-			probs.Malformed("CSR generated using a pre-1.0.2 OpenSSL with a client that doesn't properly specify the CSR version. See https://community.letsencrypt.org/t/openssl-bug-information/19591"),
-			nil,
-		)
+
+	if len(newOrderRequest.Identifiers) == 0 {
+		wfe.sendError(response, logEvent,
+			probs.Malformed("NewOrder request did not specify any identifiers"), nil)
 		return
 	}
 
-	// Check for a malformed CSR early to avoid unnecessary RPCs
-	_, err = x509.ParseCertificateRequest(rawCSR.CSR)
-	if err != nil {
-		wfe.sendError(response, logEvent, probs.Malformed("Error parsing certificate request: %s", err), err)
-		return
+	// Collect up all of the DNS identifier values into a []string for subsequent
+	// layers to process. We reject anything with a non-DNS type identifier here.
+	names := make([]string, len(newOrderRequest.Identifiers))
+	for i, ident := range newOrderRequest.Identifiers {
+		if ident.Type != core.IdentifierDNS {
+			wfe.sendError(response, logEvent,
+				probs.Malformed("NewOrder request included invalid non-DNS type identifier: type %q, value %q",
+					ident.Type, ident.Value),
+				nil)
+			return
+		}
+		names[i] = ident.Value
 	}
 
 	order, err := wfe.RA.NewOrder(ctx, &rapb.NewOrderRequest{
 		RegistrationID: &acct.ID,
-		Csr:            rawCSR.CSR,
+		Names:          names,
 	})
 	if err != nil {
 		wfe.sendError(response, logEvent, web.ProblemDetailsForError(err, "Error creating new order"), err)
 		return
 	}
 
-	respObj := orderJSON{
-		Status:         core.AcmeStatus(*order.Status),
-		Expires:        time.Unix(0, *order.Expires).UTC(),
-		CSR:            core.JSONBuffer(order.Csr),
-		Authorizations: make([]string, len(order.Authorizations)),
-	}
-	for i, authzID := range order.Authorizations {
-		respObj.Authorizations[i] = wfe.relativeEndpoint(request, fmt.Sprintf("%s%s", authzPath, authzID))
+	// Convert the corepb.Order.Names back into identifiers for display to the
```

Comment:
I think this is taken care of by orderToOrderJSON and can be deleted from this function?

Reply:
Agreed. I missed this when I made the `orderToOrderJSON` function to replace the duplicated code. Thanks!